debezium-embedded/src/test/java/io/debezium/embedded/EmbeddedEngineTest.java

```diff
@@ -328,6 +330,81 @@ public void shouldRunDebeziumEngine() throws Exception {
+    public void shouldRunEngineWithConsumerSettingOffsets() throws Exception {
+        // Add initial content to the file ...
+        appendLinesToSource(NUMBER_OF_LINES);
+
+        final Properties props = new Properties();
+        props.setProperty("name", "debezium-engine");
+        props.setProperty("connector.class", "org.apache.kafka.connect.file.FileStreamSourceConnector");
+        props.setProperty(StandaloneConfig.OFFSET_STORAGE_FILE_FILENAME_CONFIG, OFFSET_STORE_PATH.toAbsolutePath().toString());
+        props.setProperty("offset.flush.interval.ms", "0");
+        props.setProperty("file", TEST_FILE_PATH.toAbsolutePath().toString());
+        props.setProperty("topic", "topicX");
+
+        String CUSTOM_SOURCE_OFFSET_PARTITION = "test_topic_partition1";
+
+        CountDownLatch firstLatch = new CountDownLatch(1);
+        CountDownLatch allLatch = new CountDownLatch(6);
+
+        // create an engine with our custom class
+        final DebeziumEngine<RecordChangeEvent<SourceRecord>> engine = DebeziumEngine.create(ChangeEventFormat.of(Connect.class))
+                .using(props)
+                .notifying((records, committer) -> {
+                    assertThat(records.size()).isGreaterThanOrEqualTo(NUMBER_OF_LINES);
+                    Integer groupCount = records.size() / NUMBER_OF_LINES;
+
+                    for (RecordChangeEvent<SourceRecord> r : records) {
+                        DebeziumEngine.Offsets offsets = committer.buildOffsets();
+                        offsets.set(CUSTOM_SOURCE_OFFSET_PARTITION, 1L);
+                        logger.info(r.record().sourceOffset().toString());
+                        committer.markProcessed(r, offsets);
+                    }
+
+                    committer.markBatchFinished();
+                    firstLatch.countDown();
+                    for (int i = 0; i < groupCount; i++) {
+                        allLatch.countDown();
+                    }
+                })
+                .using(this.getClass().getClassLoader())
+                .build();
+
+        ExecutorService exec = Executors.newFixedThreadPool(1);
+        exec.execute(() -> {
+            LoggingContext.forConnector(getClass().getSimpleName(), "", "engine");
+            engine.run();
+        });
+
+        firstLatch.await(5000, TimeUnit.MILLISECONDS);
+        assertThat(firstLatch.getCount()).isEqualTo(0);
+
+        for (int i = 0; i < 5; i++) {
+            // Add a few more lines, and then verify they are consumed ...
+            appendLinesToSource(NUMBER_OF_LINES);
+            Thread.sleep(10);
+        }
+        allLatch.await(5000, TimeUnit.MILLISECONDS);
+        assertThat(allLatch.getCount()).isEqualTo(0);
+
+        boolean containsCustomPartition = false;
+        try (BufferedReader br = new BufferedReader(new FileReader(OFFSET_STORE_PATH.toString()))) {
+            String line;
+            while ((line = br.readLine()) != null) {
+                logger.info(line);
```

Comment:
So the line that's read back here is some binary representation, not plain text string. We should be able to get back the proper offset by reading this as a byte array and feeding it through `JsonConverter#toConnectData()` (ie. inverting the offset serialization step). While the assertion below technically passes, it feels a bit weird, and you could also do a proper lookup in the offset map then.

Reply:
Good point, updated.